project_wip_supply_cost/models/account_analytic_line.py

```diff
@@ -0,0 +1,253 @@
+        if self._is_shop_supply_account_move_reconciled():
+            raise ValidationError(_(
+                'The timesheet line {description} can not '
+                'be updated because the shop supply entry ({move_name}) is already '
+                'transfered into the cost of goods sold.'
+            ).format(
+                description=_format_timesheet_line_description(self),
+                move_name=self.shop_supply_account_move_id.name,
+            ))
+
+        self.shop_supply_account_move_id.state = 'draft'
+        vals = self._get_shop_supply_move_vals()
+        self.shop_supply_account_move_id.write(vals)
+        self.shop_supply_account_move_id.post()
+
+    def _reverse_shop_supply_account_move_for_updated_timesheet(self):
+        """Reverse the wip journal entry in the context of an updated timesheet."""
+        if self._is_shop_supply_account_move_reconciled():
+            raise ValidationError(_(
+                'The timesheet line {description} can not '
+                'be updated because the shop supply entry ({move_name}) would be '
+                'reversed. This journal entry was already transfered into '
+                'the cost of goods sold.'
+            ).format(
+                description=_format_timesheet_line_description(self),
+                move_name=self.shop_supply_account_move_id.name,
+            ))
+        self.shop_supply_account_move_id.reverse_moves()
+        self.shop_supply_account_move_id = False
+
+    def _create_update_or_reverse_shop_supply_move(self):
+        """Create / Update / Reverse the wip account move.
+
+        Depending on the status of the timesheet line,
+        the wip move is either created, updated or reversed.
+        """
+        must_create_shop_supply_move = (
+            self._requires_shop_supply_move() and not self.shop_supply_account_move_id
+        )
+        must_update_shop_supply_move = (
+            self._requires_shop_supply_move() and self.shop_supply_account_move_id
+        )
+        must_reverse_shop_supply_move = (
+            not self._requires_shop_supply_move() and self.shop_supply_account_move_id
+        )
+
+        if must_create_shop_supply_move:
+            self._create_shop_supply_move()
+
+        elif must_update_shop_supply_move:
+            self._update_shop_supply_move()
+
+        elif must_reverse_shop_supply_move:
+            self._reverse_shop_supply_account_move_for_updated_timesheet()
+
+    @api.model
+    def create(self, vals):
+        """On timesheet create, create or update the wip journal entry.
+
+        Because the creation of a timesheet is complex, the journal entry
+        may be created by a write before the return of super().create(vals).
+        """
+        line = super().create(vals)
+        if line._requires_shop_supply_move():
+            line.sudo()._create_update_or_reverse_shop_supply_move()
+        return line
+
+    def _get_shop_supply_move_dependent_fields(self):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4003316)